### ISSUE TITLE
Provide option to attack all URL parameters in DOM XSS scanner

### ISSUE THREAD

**Issue by Grunny** [zaproxy/zaproxy#2917]
I was testing an [EmberJS](http://emberjs.com/) single page app and it had a fairly straightforward XSS vulnerability in a query string that was not found by ZAP (i.e. `/path?query=<img src=x onerror=alert(0)>`). This is because it was part of a component generated client side that included data from URL parameters that are passed from the server to the client scripts. The current DOM XSS scanner only really touches issues in the hash part of the URL, however, it would be useful for testing SPAs and any other situations where client side scripts may handle query parameters to have the option to attack all query parameters similar to the normal reflected XSS scan (with all appropriate attack strings).

A simple page to test such an issue would be the following:

``` html
<!DOCTYPE html>
<html>
    <head>
        <meta charset="UTF-8">
        <title>DOM-XSS test</title>
        <script>
        function getParamValue(param) {
            var re = new RegExp('^[^#]*[&?]' + param.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&') + '=([^&#]*)' ),
                m = re.exec(document.location.href);
            if ( m && m.length > 1 ) {
                return decodeURIComponent(m[1]);
            }
            return null;
        }
        document.addEventListener('DOMContentLoaded', function () {
            document.removeEventListener('DOMContentLoaded', arguments.callee, !1);
            var queryString = getParamValue('query');
            if (queryString) {
                document.getElementById('content').innerHTML = '<p>You queried: <strong>' + queryString + '</strong></p>';
            }
        });
        </script>
    </head>
    <body>
        <div id="content">
            <form>
                <label>Query: <input name="query" type="text" /></label>
            </form>
        </div>
    </body>
</html>
```

**Comment by psiinon** [comment]
Sounds like a good plan, and thanks for the test page - thats really useful.

**Comment by kingthorin** [comment]
@psiinon @thc202 do you think it's necessary to register a new plugin ID for this?

**Comment by thc202** [comment]
No, if it's going to be changed the current scanner (that's the plan, right?).

**Comment by Grunny** [comment]
Yeah, I think it would just be to change the current scanner. I had a PoC for it working a while ago, but the DOM XSS scanner had issues opening browsers at the time so I left it.

**Comment by kingthorin** [comment]
Well I was going to add a second one based on AbstractAppParamPlugin. At least that was my original plan. (That’s what I’ve got setup so far: using the same name, I’d, etc)

**Comment by kingthorin** [comment]
I guess with a separate name/id then users have greater flexibility?

**Comment by thc202** [comment]
Fixed in zaproxy/zap-extensions#1403.

**Comment by thc202** [comment]
Released in version 7 of DOM XSS Active scanner rule add-on.
